{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5ab1552c",
   "metadata": {},
   "source": [
    "# Define functions  and classes - run these before using anything"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "eb80b578",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import copy\n",
    "import re\n",
    "import ast\n",
    "\n",
    "# Modules for obtaining SMILES strings\n",
    "from rdkit import Chem\n",
    "import time\n",
    "import pandas as pd\n",
    "import os\n",
    "import pubchempy as pcp\n",
    "import cirpy\n",
    "from urllib.error import HTTPError\n",
    "\n",
    "from py2opsin import py2opsin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "152c1e6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def my_isnull(datum) -> bool:\n",
    "    \"\"\"Take a variable and return a boolean indicating if it is a variant of null\"\"\"\n",
    "    if pd.isnull(datum) or datum == '' or datum == 'nan':\n",
    "        return True\n",
    "    try:\n",
    "        if datum.strip(',- ') == '':\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    except AttributeError:\n",
    "        return True\n",
    "    \n",
    "def clean_remarks_whitespace(text : str) -> bool:\n",
    "    \"\"\"\n",
    "    Takes a string (from the remarks section) and cleans up the text, by replacing \n",
    "    commonly erroneous motifs.\n",
    "    Returns the modified string.\n",
    "    \"\"\"\n",
    "    if my_isnull(text):\n",
    "        return text\n",
    "    else:\n",
    "        text = re.sub('\\ \\ +','  ',text)\n",
    "        text = text.replace('Conditions-not-stated', 'Conditions not stated')\n",
    "        # formatting\n",
    "        text = text.replace(\"C =\",\"C=\")\n",
    "        text = text.replace(\"C <\",\"C<\")\n",
    "        text = text.replace(\"c <\",\"c<\")\n",
    "        text = text.replace(\"I <\",\"I<\")\n",
    "        text = text.replace(\"c =\",\"c=\")\n",
    "        text = text.replace(\"C=\",\"c=\")\n",
    "        return text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3ec2b89e",
   "metadata": {},
   "outputs": [],
   "source": [
    "class DataEntryBlock(object):\n",
    "    \"\"\"\n",
    "    A container that holds multiple individual data entries.\n",
    "    The heirarchy is that a DataEntryBlock contains metadata across multiple entries, \n",
    "    such as the name of the molecule and the number in the reference book.\n",
    "    \"\"\"\n",
    "    def __init__(self, smiles=None, number=None, original_name=None, names=None, description=None, entries=None,\n",
    "                 refs=None, nicknames=None):\n",
    "        if nicknames is None:\n",
    "            nicknames = []\n",
    "        if refs is None:\n",
    "            refs = {}\n",
    "        if entries is None:\n",
    "            entries = []\n",
    "        self.smiles = smiles\n",
    "        self.number = number\n",
    "        self.original_name = original_name\n",
    "        self.refs=refs\n",
    "        self.names = names\n",
    "        self.description = description \n",
    "        self.entries = entries\n",
    "        self.nicknames = nicknames\n",
    "        \n",
    "    def __str__(self): \n",
    "        return \"# {}: {} [{} | {}], descript. {}, refs {} | {} entries: {}\\n\".format(self.number, self.smiles, self.names, self.nicknames, self.description, self.refs, len(self.entries), self.entries)\n",
    "    \n",
    "    def export_to_dict(self):\n",
    "        entries_dict = {}\n",
    "        for ct, entry in enumerate(self.entries):\n",
    "            entries_dict[ct] = entry.to_dict()\n",
    "        out_dict = {'data': {'names':{\n",
    "                                    'SMILES': self.smiles,\n",
    "                                    'IUPAC': self.names,\n",
    "                                    'nicknames': self.nicknames\n",
    "                                     },\n",
    "                             'description':self.description,\n",
    "                             'refs': self.refs,\n",
    "                             'entries':entries_dict,}\n",
    "                   }\n",
    "        return out_dict\n",
    "    \n",
    "class DataEntry(object): \n",
    "    \"\"\"\n",
    "    A block of text partially parsed from one entry # to another\n",
    "    This class provides a convenient way of storing and accessing the information.\n",
    "    This contains data for the individual pKa entry, such as the pKa value and type, temperature, method, and reference.\n",
    "    \"\"\"\n",
    "    def __init__(self, pkas=None, t=None, remarks = None, method=None, assessment=None, ref=None, subtables=None):\n",
    "        if subtables is None:\n",
    "            subtables = []\n",
    "        if ref is None:\n",
    "            ref = []\n",
    "        if assessment is None:\n",
    "            assessment = []\n",
    "        if method is None:\n",
    "            method = []\n",
    "        if t is None:\n",
    "            t = []\n",
    "        if pkas is None:\n",
    "            pkas = []\n",
    "        self.pkas = pkas\n",
    "        self.T = t\n",
    "        self.remarks = remarks\n",
    "        self.method = method\n",
    "        self.assessment = assessment\n",
    "        self.ref = ref\n",
    "        self.subtables = subtables\n",
    "\n",
    "    def __str__(self): \n",
    "        return \"pKa: {} | T: {} | Remarks: {} | Method: {} | Assessment: {} | Ref: {} \\n\".format(self.pkas, self.T, self.remarks, self.method, self.assessment, self.ref)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "34853409",
   "metadata": {},
   "outputs": [],
   "source": [
    "replacements = [\n",
    "    (\",---\", \",\"),\n",
    "    (\",--\", \",\"),\n",
    "    (\",-\", \",\"),\n",
    "    (\".--\", \",\"),\n",
    "    (\" .-\", \",\"),\n",
    "    (\" ,\", \",\"),\n",
    "    (\", .\", \",\"),\n",
    "    (\". \", \",\"),\n",
    "    (\".-\", \"-\"),\n",
    "    (\"---\", \"-\"),\n",
    "    (\"--\", \"-\"),\n",
    "    (\"’\", \"'\"),\n",
    "    (\"”\", '\"'),\n",
    "    (\"alpha\", \"α\"),\n",
    "    (\"beta\", \"β\"),\n",
    "    (\"gamma\", \"γ\"),\n",
    "    (\"Delta\", \"Δ\"),\n",
    "    (\"delta\", \"δ\")\n",
    "]\n",
    "\n",
    "\n",
    "def process_data(data : list) -> list:\n",
    "    \"\"\"\n",
    "    Turns a DataBlock and returns a list of block datas. Assumes that entry #s are always in the first column.\n",
    "    \"\"\"\n",
    "    block = DataEntryBlock(number = None, smiles=None, original_name = None, names = None, nicknames=[], description = None, entries=[],refs={})\n",
    "    block_list = []\n",
    "    for _, row in data.iterrows():\n",
    "        for index, item in enumerate(row):\n",
    "            if index == 0:  # Looking only at the first column, where we expect there to be an entry number.\n",
    "                entrynum = row[0] \n",
    "                if not pd.isnull(entrynum):\n",
    "                    # Try to clean up the IUPAC name\n",
    "                    iupacname = row[1].strip()\n",
    "                    for replacement in replacements:\n",
    "                        iupacname = iupacname.replace(replacement[0], replacement[1])\n",
    "\n",
    "                    # Turn the entrynumber into an integer, if possible\n",
    "                    try:\n",
    "                        entrynum = int(entrynum)\n",
    "                    except ValueError:\n",
    "                        pass\n",
    "                    block_list.append(block) # Add previous block because it's done parsing\n",
    "                    block = DataEntryBlock(number = entrynum, smiles=None, original_name = iupacname, names = None, nicknames = [], description = None, entries=[],refs={})\n",
    "                    description = row[5]\n",
    "                    if not pd.isnull(description):\n",
    "                        block.description = \" - \" + description\n",
    "                        print(row[5])\n",
    "                else:\n",
    "                    row = row.drop(['Entry #','IUPAC'])\n",
    "                    block.entries.append(row)\n",
    "\n",
    "    block_list.append(block) # Add last block\n",
    "    block_list.remove(block_list[0])# Remove first block\n",
    "    return block_list\n",
    "\n",
    "def sort_block_list(block_list : list) -> list:\n",
    "    \"\"\"\n",
    "    Sort a block list by the entry numbers. Returns the sorted blocklist.\n",
    "    \"\"\"\n",
    "    new_block_list = []\n",
    "    block_dict = {}\n",
    "    for block in block_list:\n",
    "        block_dict[str(block.number)] = block\n",
    "    sortedkeys = sorted(block_dict, key=str.lower)\n",
    "    for key in sortedkeys:\n",
    "        new_block_list.append(block_dict[key])\n",
    "    return new_block_list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eccb0adc",
   "metadata": {},
   "source": [
    "### Load in .csvs and concatenate them into 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6918c335",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"test_sample_out.csv\")\n",
    "dfs = {\"1\": df}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "74ce79b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort through the keys and reorder them when concatenating\n",
    "sortedkeys = sorted(dfs, key=str.lower)\n",
    "#print(sortedkeys)\n",
    "df_cat = pd.DataFrame()\n",
    "for key in sortedkeys:\n",
    "    df = dfs[key]\n",
    "    df_cat = pd.concat([df_cat, df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "274deecc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry #</th>\n",
       "      <th>IUPAC</th>\n",
       "      <th>pKa type</th>\n",
       "      <th>pKa</th>\n",
       "      <th>T</th>\n",
       "      <th>Remarks</th>\n",
       "      <th>Method</th>\n",
       "      <th>Assessment</th>\n",
       "      <th>Ref</th>\n",
       "      <th>(Subtable)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2004.0</td>\n",
       "      <td>Methanol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>15.50</td>\n",
       "      <td>25.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>B8</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>15.09</td>\n",
       "      <td>25.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KIN</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>M126</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2005.0</td>\n",
       "      <td>Methyl hydroperoxide</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>11.50</td>\n",
       "      <td>20.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O5</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>E27</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2006.0</td>\n",
       "      <td>Methanethiol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>10.33</td>\n",
       "      <td>25.0</td>\n",
       "      <td>1% ethanol  gas solubility method</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>K57</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2007.0</td>\n",
       "      <td>Methane, trinitro-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>0.14</td>\n",
       "      <td>20.0</td>\n",
       "      <td>In aqueous HClO4 Ho scale</td>\n",
       "      <td>O6</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>T58a</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>25.5</td>\n",
       "      <td>Mixed constant</td>\n",
       "      <td>O5</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>H5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>0.05</td>\n",
       "      <td>9.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>0.23</td>\n",
       "      <td>5.0</td>\n",
       "      <td>In aqueous HCl Ho scale</td>\n",
       "      <td>O6</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>N39, S82</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>0.17</td>\n",
       "      <td>20.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>0.11</td>\n",
       "      <td>40.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60.0</td>\n",
       "      <td>0.02  Thermodynamic quantities are derived fro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2008.0</td>\n",
       "      <td>Methane, dinitro-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>3.63</td>\n",
       "      <td>20.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O5</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>N39</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>3.72</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O5</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>N39</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>3.60</td>\n",
       "      <td>20.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>3.51</td>\n",
       "      <td>40.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>3.43</td>\n",
       "      <td>60.0</td>\n",
       "      <td>Thermodynamic quantities are derived from the ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>3.57</td>\n",
       "      <td>25.0</td>\n",
       "      <td>I=0.06</td>\n",
       "      <td>O5</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>A7</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>2009.0</td>\n",
       "      <td>Methane, nitro-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>10.45</td>\n",
       "      <td>10.0</td>\n",
       "      <td>C=0.005-0.017  mixed constant</td>\n",
       "      <td>E3bg</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>T67</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>10.33</td>\n",
       "      <td>18.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>10.21</td>\n",
       "      <td>25.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>10.24</td>\n",
       "      <td>25.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E3bg</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>W22</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>2010.0</td>\n",
       "      <td>Methanohydroxamic acid (Formohydroxamic acid)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>8.65</td>\n",
       "      <td>NaN</td>\n",
       "      <td>I=0.2(NaCl)</td>\n",
       "      <td>E3bg</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>C72</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>2011.0</td>\n",
       "      <td>Methanedithioic acid, amino- (Dithiocarbamic a...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>2.95</td>\n",
       "      <td>25.0</td>\n",
       "      <td>C=0.002-0.01</td>\n",
       "      <td>C2</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>G4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>2012.0</td>\n",
       "      <td>Methanesulfinic acid, hydroxy-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__0</td>\n",
       "      <td>1.65</td>\n",
       "      <td>20.0</td>\n",
       "      <td>C=0.1  mixed constant</td>\n",
       "      <td>E3bg</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>R54</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Entry #                                              IUPAC pKa type  \\\n",
       "0    2004.0                                           Methanol      NaN   \n",
       "1       NaN                                                NaN      __0   \n",
       "2       NaN                                                NaN      __0   \n",
       "3    2005.0                               Methyl hydroperoxide      NaN   \n",
       "4       NaN                                                NaN      __0   \n",
       "5    2006.0                                       Methanethiol      NaN   \n",
       "6       NaN                                                NaN      __0   \n",
       "7    2007.0                                 Methane, trinitro-      NaN   \n",
       "8       NaN                                                NaN      __0   \n",
       "9       NaN                                                NaN      __0   \n",
       "10      NaN                                                NaN      __0   \n",
       "11      NaN                                                NaN      __0   \n",
       "12      NaN                                                NaN      __0   \n",
       "13      NaN                                                NaN      __0   \n",
       "14      NaN                                                NaN      NaN   \n",
       "15   2008.0                                  Methane, dinitro-      NaN   \n",
       "16      NaN                                                NaN      __0   \n",
       "17      NaN                                                NaN      __0   \n",
       "18      NaN                                                NaN      __0   \n",
       "19      NaN                                                NaN      __0   \n",
       "20      NaN                                                NaN      __0   \n",
       "21      NaN                                                NaN      __0   \n",
       "22   2009.0                                    Methane, nitro-      NaN   \n",
       "23      NaN                                                NaN      __0   \n",
       "24      NaN                                                NaN      __0   \n",
       "25      NaN                                                NaN      __0   \n",
       "26      NaN                                                NaN      __0   \n",
       "27   2010.0      Methanohydroxamic acid (Formohydroxamic acid)      NaN   \n",
       "28      NaN                                                NaN      __0   \n",
       "29   2011.0  Methanedithioic acid, amino- (Dithiocarbamic a...      NaN   \n",
       "30      NaN                                                NaN      __0   \n",
       "31   2012.0                     Methanesulfinic acid, hydroxy-      NaN   \n",
       "32      NaN                                                NaN      __0   \n",
       "\n",
       "      pKa     T                                            Remarks Method  \\\n",
       "0     NaN   NaN                                                NaN    NaN   \n",
       "1   15.50  25.0                                                NaN     C3   \n",
       "2   15.09  25.0                                                NaN    KIN   \n",
       "3     NaN   NaN                                                NaN    NaN   \n",
       "4   11.50  20.0                                                NaN     O5   \n",
       "5     NaN   NaN                                                NaN    NaN   \n",
       "6   10.33  25.0                  1% ethanol  gas solubility method    NaN   \n",
       "7     NaN   NaN                                                NaN    NaN   \n",
       "8    0.14  20.0                          In aqueous HClO4 Ho scale     O6   \n",
       "9    0.06  25.5                                     Mixed constant     O5   \n",
       "10   0.05   9.6                                                NaN    NaN   \n",
       "11   0.23   5.0                            In aqueous HCl Ho scale     O6   \n",
       "12   0.17  20.0                                                NaN    NaN   \n",
       "13   0.11  40.0                                                NaN    NaN   \n",
       "14    NaN  60.0  0.02  Thermodynamic quantities are derived fro...    NaN   \n",
       "15    NaN   NaN                                                NaN    NaN   \n",
       "16   3.63  20.0                                                NaN     O5   \n",
       "17   3.72   5.0                                                NaN     O5   \n",
       "18   3.60  20.0                                                NaN    NaN   \n",
       "19   3.51  40.0                                                NaN    NaN   \n",
       "20   3.43  60.0  Thermodynamic quantities are derived from the ...    NaN   \n",
       "21   3.57  25.0                                             I=0.06     O5   \n",
       "22    NaN   NaN                                                NaN    NaN   \n",
       "23  10.45  10.0                      C=0.005-0.017  mixed constant   E3bg   \n",
       "24  10.33  18.0                                                NaN    NaN   \n",
       "25  10.21  25.0                                                NaN    NaN   \n",
       "26  10.24  25.0                                                NaN   E3bg   \n",
       "27    NaN   NaN                                                NaN    NaN   \n",
       "28   8.65   NaN                                        I=0.2(NaCl)   E3bg   \n",
       "29    NaN   NaN                                                NaN    NaN   \n",
       "30   2.95  25.0                                       C=0.002-0.01     C2   \n",
       "31    NaN   NaN                                                NaN    NaN   \n",
       "32   1.65  20.0                              C=0.1  mixed constant   E3bg   \n",
       "\n",
       "   Assessment       Ref  (Subtable)  \n",
       "0         NaN       NaN         NaN  \n",
       "1     Uncert.        B8         NaN  \n",
       "2     Uncert.      M126         NaN  \n",
       "3         NaN       NaN         NaN  \n",
       "4     Uncert.       E27         NaN  \n",
       "5         NaN       NaN         NaN  \n",
       "6     Uncert.       K57         NaN  \n",
       "7         NaN       NaN         NaN  \n",
       "8     Uncert.      T58a         NaN  \n",
       "9     Uncert.        H5         NaN  \n",
       "10        NaN       NaN         NaN  \n",
       "11    Uncert.  N39, S82         NaN  \n",
       "12        NaN       NaN         NaN  \n",
       "13        NaN       NaN         NaN  \n",
       "14        NaN       NaN         NaN  \n",
       "15        NaN       NaN         NaN  \n",
       "16    Approx.       N39         NaN  \n",
       "17    Approx.       N39         NaN  \n",
       "18        NaN       NaN         NaN  \n",
       "19        NaN       NaN         NaN  \n",
       "20        NaN       NaN         NaN  \n",
       "21    Approx.        A7         NaN  \n",
       "22        NaN       NaN         NaN  \n",
       "23    Approx.       T67         NaN  \n",
       "24        NaN       NaN         NaN  \n",
       "25        NaN       NaN         NaN  \n",
       "26    Approx.       W22         NaN  \n",
       "27        NaN       NaN         NaN  \n",
       "28    Uncert.       C72         NaN  \n",
       "29        NaN       NaN         NaN  \n",
       "30    Approx.        G4         NaN  \n",
       "31        NaN       NaN         NaN  \n",
       "32    Approx.       R54         NaN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cat.head(999)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "57efe0b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cat.to_csv(\"sample_all_concat.csv\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a328a093",
   "metadata": {},
   "source": [
    "### Convert and sort blocklist from .csv data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d09d4cb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:35: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  iupacname = row[1].strip()\n",
      "/tmp/ipykernel_187751/2694344708.py:46: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  description = row[5]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:35: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  iupacname = row[1].strip()\n",
      "/tmp/ipykernel_187751/2694344708.py:46: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  description = row[5]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:35: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  iupacname = row[1].strip()\n",
      "/tmp/ipykernel_187751/2694344708.py:46: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  description = row[5]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:35: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  iupacname = row[1].strip()\n",
      "/tmp/ipykernel_187751/2694344708.py:46: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  description = row[5]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:35: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  iupacname = row[1].strip()\n",
      "/tmp/ipykernel_187751/2694344708.py:46: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  description = row[5]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:35: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  iupacname = row[1].strip()\n",
      "/tmp/ipykernel_187751/2694344708.py:46: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  description = row[5]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:35: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  iupacname = row[1].strip()\n",
      "/tmp/ipykernel_187751/2694344708.py:46: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  description = row[5]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:35: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  iupacname = row[1].strip()\n",
      "/tmp/ipykernel_187751/2694344708.py:46: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  description = row[5]\n",
      "/tmp/ipykernel_187751/2694344708.py:32: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  entrynum = row[0]\n",
      "/tmp/ipykernel_187751/2694344708.py:35: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  iupacname = row[1].strip()\n",
      "/tmp/ipykernel_187751/2694344708.py:46: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  description = row[5]\n"
     ]
    }
   ],
   "source": [
    "blocklist = process_data(df_cat)\n",
    "sorted_blocklist = sort_block_list(blocklist)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "397df014",
   "metadata": {},
   "source": [
    "### Test if every number is represented\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e687f46d",
   "metadata": {},
   "outputs": [],
   "source": [
    "blocknums = []\n",
    "for block in sorted_blocklist:\n",
    "    blocknums.append(block.number)\n",
    "\n",
    "allnums = list(range(2004,2012)) # manually-determined first and last indices from the dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a4e2a4d-1ba6-4b36-bf53-1f77827244c7",
   "metadata": {},
   "source": [
    "This should be empty if we've got all the entries represented."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "caef7456",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(set(allnums) - set(blocknums))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1ed75de",
   "metadata": {},
   "source": [
    "## Postprocess"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c817c33d",
   "metadata": {},
   "source": [
    "### Remove comments that only have remarks. In those cases, add remarks to block description."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "28942cfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "for block in sorted_blocklist:\n",
    "    for count, entry in enumerate(block.entries):\n",
    "        if my_isnull(entry['pKa']) and my_isnull(entry['pKa type']) and my_isnull(entry['T']) and my_isnull(entry['Method']) and my_isnull(entry['Ref']) and my_isnull(entry['Assessment']) and not my_isnull(entry['Remarks']):\n",
    "            if block.description is None:\n",
    "                block.description = \" - \" + entry['Remarks']\n",
    "            else:\n",
    "                block.description = block.description + \"\\n - \" + entry['Remarks']\n",
    "            block.entries.pop(count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "859a65e2",
   "metadata": {},
   "source": [
    "### Replace many whitespaces with just two and clean up some stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "543e17d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "for block in sorted_blocklist:\n",
    "    for count, e in enumerate(block.entries):\n",
    "        block.entries[count]['Remarks'] = clean_remarks_whitespace(e['Remarks'])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25fc584f",
   "metadata": {},
   "source": [
    "### Deconvolute IUPAC names into ensemble of names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "64a9c9a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "WORD_PATTERN = r\"[a-z0-9\\-,A-Z\\(\\)']*-?[\\.A-MO-Z][0-9\\-a-z']+[\\[\\]\\.\\'\\\"\\’\\”HNOa-z,\\-0-9\\(\\)]*(?: acid| chloride| bromide| dichloride| fluoborate| acetate| iodide| oxime| dihydrogen phosphate| dihydrogen phosphorothioate| phosphoramidate| dioxime| thiocyanate)?\\)?\"\n",
    "WORD_PATTERN_WITH_N = r\"[a-z0-9\\-,A-Z\\(\\)']*-?[\\.A-Z][0-9\\-a-z']+[\\[\\]\\.\\'\\\"HNOa-z,\\-0-9\\(\\)]*(?: acid| chloride| bromide| dichloride| fluoborate| acetate| iodide| oxime| dihydrogen phosphate| dihydrogen phosphorothioate| phosphoramidate| dioxime| thiocyanate)?\\)?\"\n",
    "NICKNAME_PATTERN = r\"\\ \\([,-_\\w\\ 0-9']*\\(?[,-_\\w\\ 0-9]*\\)?[\\\"’,'-_\\w0-9]*\\)[\\-]?$\"\n",
    "def lowercase_func(s : str, ignore_n = True) -> str:\n",
    "    \"\"\"\n",
    "    Convert a string into its lowercase form to standardize it. But, don't do this for certain motifs.\n",
    "    \"\"\"\n",
    "    retstring = ''\n",
    "    for ct, c in enumerate(s):\n",
    "        if ct == 0 and (c != 'N' or ignore_n == False):\n",
    "            retstring += c.lower() \n",
    "        else:\n",
    "            retstring += c\n",
    "            \n",
    "    # special cases: \"O,O\"\n",
    "    retstring = retstring.replace(\"o,O\",\"O,O\")\n",
    "    retstring = retstring.replace(\"n,N\",\"N,N\")\n",
    "    return retstring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "230484f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_nickname(name : str, debug=False) -> (str, list):\n",
    "    \"\"\"\n",
    "    Some entries have both a name and a nickname like Acetone (Propanone). \n",
    "    This method detects those names and splits them up, and returns a tuple of the name and nickname.\n",
    "    Nicknames is a list because multiple nicknames can be found. Name is a string.\n",
    "    \"\"\"\n",
    "    # special case: name starts with parentheses\n",
    "    starts_with_parentheses = False\n",
    "    if name[0] == '(':\n",
    "        starts_with_parentheses = True\n",
    "    match = re.findall(NICKNAME_PATTERN,name)\n",
    "    if len(match) > 0:\n",
    "        nicknames = [i.strip(' ') for i in match]\n",
    "        for j in nicknames: # remove the matching nickname from the full name\n",
    "            name = name.replace(j,'').strip('()').rstrip(' ')\n",
    "#            print(\"Nickname : {}\".format(j))\n",
    "#            print(name)\n",
    "        nicknames = [fix_nickname_parentheses(i.strip('( )-')) for i in nicknames]    \n",
    "        if starts_with_parentheses:\n",
    "            name = '(' + name\n",
    "    else:\n",
    "        nicknames = []\n",
    "    if debug:\n",
    "        print(match)\n",
    "    return name, nicknames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7c62eda9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fix_nickname_parentheses(nickname : str) -> str:\n",
    "    \"\"\"\n",
    "    Takes a nickname and fixes it if the parentheses got messed up. \n",
    "    Returns the corrected nickname.\n",
    "    \"\"\"\n",
    "    if nickname.count('(') > nickname.count(')'):\n",
    "        nickname = nickname + ')'\n",
    "    if nickname.count('(') < nickname.count(')'):\n",
    "        nickname = '(' + nickname\n",
    "    return nickname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d140f7a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is a list of corrections if there are any molecules that need to be manually corrected.\n",
    "MANUAL_NAMES = ['Desferriferrioxamin B, N-acetyl-', 'Uridylic polynucleotide, 5-bromo-',\n",
    "               ]\n",
    "\n",
    "NICKNAMES_DICT = {}\n",
    "\n",
    "#\n",
    "\n",
    "SPECIAL_NAMES_LIST = ['ketone','phosphate','acid','trioxime','phosphorodithioate',\n",
    "                      'cation','hydrobromide','ester','dioxide', 'Phenyl',\n",
    "                      'Benzyl', 'O-Butyl', 'O-Isopropyl', 'O-Isobutyl',\n",
    "                      'chloride','Red','Purple','deoxystreptamine', \n",
    "                      'Methy', 'Ethyl', 'phosphonothioate', \"Diethyl\",\n",
    "                      'Hydroxyphenyl', 'Chlorobenzaldehyde',\n",
    "                      'hydroperoxide','dioxime', \n",
    "                      'isobutylphosphonothioate']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0c7ce8d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_name(name : str, debug=False):\n",
    "    \"\"\"\n",
    "    Given a name, extract it into a name and nicknames.\n",
    "    For the name, try to determine the name 'backbone' or the leading name,\n",
    "    and also the descriptor attached to the leading name.\n",
    "    Returns the cleaned-up names and nicknames.\n",
    "    \"\"\"\n",
    "    names = []\n",
    "\n",
    "    # for a few species, we need to check if its both tricky to parse the nickname AND FUll name\n",
    "    if name in NICKNAMES_DICT.keys():\n",
    "        names = NICKNAMES_DICT[name]['Names']\n",
    "        nicknames = NICKNAMES_DICT[name]['Nicknames']\n",
    "        return names, nicknames\n",
    "    \n",
    "    \n",
    "    # First, just search for nicknames. e.g. Methanal (Formaldehyde) --> Formaldehyde\n",
    "    name, nicknames = find_nickname(name)\n",
    "    if len(nicknames) > 0:\n",
    "        name, nicknames_append = find_nickname(name)\n",
    "        if nicknames_append:\n",
    "            nicknames.append(nicknames_append[0])\n",
    "            name, nicknames_append = find_nickname(name)\n",
    "            if nicknames_append:\n",
    "                nicknames.append(nicknames_append[0])\n",
    "            \n",
    "    # check if it's in the Manual Names list\n",
    "    if name in MANUAL_NAMES:\n",
    "        names = [name]\n",
    "        return names, nicknames\n",
    "\n",
    "    split_words = name.split()\n",
    "    # If we just have 1 entry, or it's a \"reference\" entry, then no need to postprocess.\n",
    "    if len(split_words) == 1 or (\"see\" in split_words and \"For\" in split_words):\n",
    "        names.append(name)\n",
    "        #print(name)\n",
    "\n",
    "        \n",
    "    # simple cases e.g. \"Sodium chloride\" shouldn't be decomposed\n",
    "    elif (len(split_words) >= 2 and (',' not in name and '-' not in name)) or (any([x in name for x in SPECIAL_NAMES_LIST]) and (', ' not in name and name[-1] != ',')):\n",
    "        names.append(name)\n",
    "#        print(name)\n",
    "    # Otherwise... Try to identify the name as a format of [Main chemical name,] + [descriptor-]\n",
    "    else:\n",
    "        if \"Naphth\" in name or \"Nonane\" in name:\n",
    "            match = re.search(WORD_PATTERN_WITH_N,name) \n",
    "            ignore_n = False\n",
    "        else:\n",
    "            match = re.search(WORD_PATTERN,name) \n",
    "            ignore_n = True\n",
    "        if match is not None: # just take the first name as this can lead to multiple matches\n",
    "            leading_name = match.group().strip(', -')\n",
    "            substring = name.replace(leading_name,'').strip(', ')\n",
    "            if len(substring.replace(' ','')) > 0: # basically if it's not just blanks\n",
    "            #print(leading_name + ' | ' + substring + ' | ' + block.name.strip())\n",
    "            \n",
    "            # Permutation one: [descriptor-main_chemical_name]\n",
    "            \n",
    "                descriptor_name = substring.rstrip('-')+lowercase_func(leading_name, ignore_n = ignore_n).rstrip(',')\n",
    "                descriptor_name_dash = substring.rstrip('-')+'-'+lowercase_func(leading_name, ignore_n = ignore_n).rstrip(',')\n",
    "                original_name = lowercase_func(leading_name, ignore_n = ignore_n).rstrip(',') + ', ' + substring.rstrip('-')+'-'\n",
    "                names.append(original_name)\n",
    "                if leading_name[0].isdigit():\n",
    "                    names.append(descriptor_name_dash)\n",
    "                else:\n",
    "                    names.append(descriptor_name)\n",
    "            else:\n",
    "                names.append(leading_name.rstrip(','))\n",
    "            if debug:\n",
    "                print('Lead: '+ leading_name)\n",
    "                print('Substring: '+substring)\n",
    "        else:\n",
    "            names.append(name)\n",
    "    \n",
    "    return names, nicknames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1be4f41f",
   "metadata": {},
   "outputs": [],
   "source": [
    "for block in sorted_blocklist:\n",
    "    name = block.original_name\n",
    "    names, nicknames = process_name(name)\n",
    "    block.names = names\n",
    "    block.nicknames = nicknames"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2b099ed",
   "metadata": {},
   "source": [
    "### Copy methods/other data if no reference for entry. Also, add \"source-specific\" comments to the reference description e.g. Thermodynamic quantities are derived from the results or Other measurements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9d65a659",
   "metadata": {},
   "outputs": [],
   "source": [
    "OTHER_PATTERNS = r'Other [a-z]*:[ A-Za-z0-9,]*[.]?'\n",
    "OTHER_PATTERNS_INCLUSIVE = r'Other [a-z ]* in [ A-Za-z0-9,]*[.]?'\n",
    "FOR_PATTERNS = r'For [ \\[\\]a-z-A-Z0-9\\'.\\,]*see[ \\[\\]a-zA-Z0-9\\'.]*([A-Z]*[0-9]*[a-z]*[, ]?)*[.]?'\n",
    "REF_COMMENTS = ['Thermodynamic quantities for the keto and enol forms are derived from the results',\n",
    "                'Thermodynamic quantities are derived from the results for both enol and keto forms',\n",
    "                'Thermodynamic quantities are derived from the results', \n",
    "                'Thermodynamic quantities are also given',\n",
    "                'Thermodynamic data are also given',\n",
    "                'Thermodynamic quantities also given',\n",
    "                'Thermodynamic values are derived from the results'\n",
    "                'Thermodynamic quantities are derived from these results',\n",
    "                'Thermodynamic quantities are derived from the results for both enol and keto forms',\n",
    "                'Thermodynamic quantities are given',\n",
    "                'Values in other inert salt solutions are also given',\n",
    "                'Values in mixed solvents are also given',\n",
    "                'pK assignment discussed',\n",
    "                'Value in mixed solvent is also given',\n",
    "                'Thermodynamic quantities are derived from the results.2-Azapropane-1,3-diphosphonic acid, 2-ethyl- (Iminodi(methylphosphonic acid) N-ethyl-)'\n",
    "                \n",
    "                \n",
    "                \n",
    "               ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e2092cf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "for block in sorted_blocklist:\n",
    "    # Additional cleaning. \n",
    "    # This transforms the literal reference book text into metadata.\n",
    "    # For example, if we have to infer the temperature from the previous line, then we copy the previous line's temperature\n",
    "    # into the new entry's T. (This is because sometimes, for example, an entry will have the same exact metadata from one line\n",
    "    # to the next, which is inferred because its metadata is blank). \n",
    "    for count, entry in enumerate(block.entries):\n",
    "        \n",
    "        ## ONLY TAG THE REFS\n",
    "        if my_isnull(entry['Ref']):\n",
    "            block.entries[count]['Ref'] = block.entries[count-1]['Ref']\n",
    "            # Check for other missing stuff.\n",
    "            if my_isnull(entry['T']) and not my_isnull(block.entries[count-1]['T']):\n",
    "                if '(assumed)' in str(block.entries[count-1]['T']):\n",
    "                    add_string = ''\n",
    "                else:\n",
    "                    add_string = \" (assumed)\"\n",
    "                entry['T'] = str(block.entries[count-1]['T']) + add_string\n",
    "            if my_isnull(entry['Method']):\n",
    "                entry['Method'] = block.entries[count-1]['Method']\n",
    "            if my_isnull(entry['Assessment']):\n",
    "                entry['Assessment'] = block.entries[count-1]['Assessment'] \n",
    "\n",
    "        else:\n",
    "            block.refs[entry['Ref']] = []\n",
    "\n",
    "        # Move remarks from the entry's remarks to the BLOCK's remarks (for metadata that\n",
    "        #     pervades to the entire molecule, not just the line of pKa data.\n",
    "        if not my_isnull(entry['Remarks']):\n",
    "            for ref_comment in REF_COMMENTS:\n",
    "                if ref_comment in entry['Remarks']:\n",
    "                    try:\n",
    "                        block.refs[entry['Ref']].append(ref_comment)\n",
    "                    except KeyError:\n",
    "                        block.refs[entry['Ref']] = [ref_comment]\n",
    "                    entry['Remarks'] = entry['Remarks'].replace(ref_comment,'')\n",
    "            # Same as above, but for other special patterns.\n",
    "            for pattern in [OTHER_PATTERNS, OTHER_PATTERNS_INCLUSIVE, FOR_PATTERNS]:\n",
    "                match = re.search(pattern,entry['Remarks'])\n",
    "                if match is not None:\n",
    "                    try:\n",
    "                        block.refs[entry['Ref']].append(match.group())\n",
    "                    except KeyError:\n",
    "                        block.refs[entry['Ref']] = match.group()\n",
    "                    entry['Remarks'] = entry['Remarks'].replace(match.group(),'')\n",
    "                    print(str(block.number) + \" \" + str(block.refs))            \n",
    "                    print(str(block.refs))\n",
    "                    \n",
    "        # NOW MODIFY THE ENTRIES\n",
    "        # Check for other missing stuff.\n",
    "        if my_isnull(entry['Remarks']) or entry['Remarks'] == '.' or entry['Remarks'] == '. ':\n",
    "            entry['Remarks'] = block.entries[count-1]['Remarks']\n",
    "\n",
    "# Iterate one more time. If the thermo remarks are in the remarks, then drop it\n",
    "# If thermo is in the block.description, then move it to the block's LAST ref.\n",
    "for block in sorted_blocklist:\n",
    "    for count, entry in enumerate(block.entries):\n",
    "        if not my_isnull(entry['Remarks']):\n",
    "            for ref_comment in REF_COMMENTS:\n",
    "                # check if in remarks\n",
    "                if ref_comment in entry['Remarks']:\n",
    "                    block.entries[count]['Remarks'] = entry['Remarks'].replace(ref_comment,'')\n",
    "\n",
    "    # check if there are reference comments in description; if so, move them to the refs rather than description\n",
    "    if block.description:\n",
    "        for ref_comment in REF_COMMENTS:\n",
    "            if ref_comment in block.description:\n",
    "                ref_to_change = list(block.refs)[-1]\n",
    "                block.refs[ref_to_change].append(ref_comment)\n",
    "                block.description = block.description.replace(' - '+ref_comment, '')\n",
    "            \n",
    "    # Lastly, if there's \"Other measurements\" in an entry's comments, then move it to the DataEntryBlock\n",
    "    for ref in block.refs:\n",
    "        if len(block.refs[ref]) > 0:\n",
    "            for ct3, content in enumerate(block.refs[ref]):\n",
    "                for pattern in [OTHER_PATTERNS, OTHER_PATTERNS_INCLUSIVE, FOR_PATTERNS]:\n",
    "                    match = re.search(pattern,content)\n",
    "                    if match is not None:\n",
    "                        block.refs[ref][ct3] = content.replace(match.group(), '')\n",
    "                        if block.description is None:\n",
    "                            block.description = \" - \" + match.group()\n",
    "                        else:\n",
    "                            block.description = block.description + \"\\n - \" + match.group()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38210cc7",
   "metadata": {},
   "source": [
    "### If many pKas for an entry, separate them into separate pKas. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "de8813ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "for block in sorted_blocklist:\n",
    "    for count, entry in enumerate(block.entries):\n",
    "        entry_copy = entry.copy()\n",
    "        try:\n",
    "            pkas = str(entry_copy['pKa']).split(',')\n",
    "            pka_types = str(entry_copy['pKa type']).split(',')\n",
    "            for ct, pka_type in enumerate(pka_types):\n",
    "                if '__0' in pka_type:\n",
    "                    pka_types[ct] = 'pKa'\n",
    "                    if len(pka_types) == 1:\n",
    "                        block.entries[count]['pKa type'] = 'pKa'\n",
    "            if len(pkas) != len(pka_types):\n",
    "                print(\"Erroneous pKa tabulation in entry # {}\".format(block.number))\n",
    "            if len(pkas) > 1:\n",
    "                entry_list = []\n",
    "                for num, pka in enumerate(pkas):\n",
    "                    duplicate = entry.copy()\n",
    "                    duplicate['pKa type'] = pka_types[num].strip()\n",
    "                    duplicate['pKa'] = pka.strip()\n",
    "                    entry_list.append(duplicate)\n",
    "                block.entries[count:count+len(pkas)-1] = entry_list\n",
    "            else:\n",
    "                pass\n",
    "        except AttributeError:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d3be2d9",
   "metadata": {},
   "source": [
    "### Clean entry comments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "dc761ee4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nan\n",
      "nan\n",
      "nan\n",
      "1% ethanol  gas solubility method\n",
      "In aqueous HClO4 Ho scale\n",
      "Mixed constant\n",
      "Mixed constant\n",
      "In aqueous HCl Ho scale\n",
      "In aqueous HCl Ho scale\n",
      "In aqueous HCl Ho scale\n",
      "I=0.06\n",
      "I=0.06\n",
      "I=0.06\n",
      "I=0.06\n",
      "I=0.06\n",
      "I=0.06\n",
      "c=0.005-0.017  mixed constant\n",
      "c=0.005-0.017  mixed constant\n",
      "c=0.005-0.017  mixed constant\n",
      "c=0.005-0.017  mixed constant\n",
      "I=0.2(NaCl)\n",
      "c=0.002-0.01\n",
      "c=0.1  mixed constant\n"
     ]
    }
   ],
   "source": [
    "for block in sorted_blocklist:\n",
    "    for count, entry in enumerate(block.entries):\n",
    "        try:\n",
    "            remarks = str(entry['Remarks'])\n",
    "            print(remarks)\n",
    "            if remarks.strip() == '.' or remarks.strip() == '-':\n",
    "                block.entries[count]['Remarks'] = ''\n",
    "            else:\n",
    "                block.entries[count]['Remarks'] = block.entries[count]['Remarks'].replace(\"  \", \", \")\n",
    "        except AttributeError:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3761eb7a",
   "metadata": {},
   "source": [
    "##  Tool for adding subtables (Not used for this demo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "cfcc1f0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_remarks(entry_duplicate, x, running_t, running_p, running_i, P_units):\n",
    "    \"\"\"\n",
    "    Given an entry and booleans representing the independent variable on the subtable,\n",
    "    format the entry's remarks so that it fits the subtable format.\n",
    "    Returns the updated entry.\n",
    "    \"\"\"\n",
    "    # Case 1: Temperature\n",
    "    if running_t:\n",
    "        entry_duplicate['T'] = x\n",
    "    elif running_p:\n",
    "        if my_isnull(entry_duplicate['Remarks']):\n",
    "            entry_duplicate['Remarks'] = 'P=' + x + ' ({})'.format(P_units)\n",
    "        else:\n",
    "            entry_duplicate['Remarks'] += '\\nP=' + x + ' ({})'.format(P_units)\n",
    "    elif running_i:\n",
    "        if my_isnull(entry_duplicate['Remarks']):\n",
    "            entry_duplicate['Remarks'] = 'I=' + str(x)\n",
    "        else:\n",
    "            entry_duplicate['Remarks'] += '\\nI=' + str(x)\n",
    "            \n",
    "    return entry_duplicate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "603d4571",
   "metadata": {},
   "outputs": [],
   "source": [
    "# In some entries, there are special \"lambda\" values that are processed differently.\n",
    "lambdas = ['275.89', '305.32', '334.77', '364.06', '421.97', '450.21', '478.10', '505.48']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0e0a94cf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def check_subtable_type(variation_header):\n",
    "    \"\"\"\n",
    "    Given a line that represents the header of a subtable,\n",
    "    returns the booleans corresponding to the type of subtable (i.e. whether the variation\n",
    "    is over temp, pressure, ionic strength, etc.) Also returns P units ('' if not specified).\n",
    "    \"\"\"\n",
    "    P_units = \"\"\n",
    "    if 'with temperature' in variation_header:\n",
    "        running_t = True\n",
    "    elif 'pressure' in variation_header:\n",
    "        running_p = True\n",
    "        P_units = variation_header.replace('Variation with pressure','').strip('\\(\\)')\n",
    "    elif 'with I' in variation_header or 'ionic strength' in variation_header:\n",
    "        running_i = True\n",
    "    else:\n",
    "        raise ValueError(\"unknown x variable type\")\n",
    "\n",
    "    return running_t, running_p, running_i, P_units\n",
    "\n",
    "def process_subtable(subtable_as_list):\n",
    "    \"\"\"\n",
    "    Takes a subtable (presented as a list), and returns:\n",
    "    - the x variable that is being varied\n",
    "    - the \"end dict\" that pairs the different pKas across the x variables\n",
    "    y dict: {'pk1': [1, 2, 3], 'pk2': [4, 5, 6]}\n",
    "    \"\"\"\n",
    "    end_dict = {}\n",
    "    x_var_list = []\n",
    "    for i in range(1,len(subtable_as_list)):\n",
    "        row = subtable_as_list[i]\n",
    "        if type(row) == str:\n",
    "            row = row.split()\n",
    "        if type(row) == list: # This case defines the x variable, e.g. pressure or temperature                \n",
    "            x_var_list.append(row)\n",
    "            \n",
    "        elif type(row) == dict: # This case is for the pkas\n",
    "            t = list(row.items())\n",
    "            archetype = t[0][0]\n",
    "            list_to_add = []\n",
    "            for item in t:\n",
    "                list_to_add.append(item[1])\n",
    "            end_dict[archetype] = list_to_add\n",
    "            \n",
    "        else:\n",
    "            raise ValueError(\"unknown row type: {}\".format(row))    \n",
    "\n",
    "    return x_var_list, end_dict\n",
    "\n",
    "processed_string='already_processed'\n",
    "\n",
    "for block in sorted_blocklist:\n",
    "    for count, entry in enumerate(block.entries):\n",
    "        if not my_isnull(entry['(Subtable)']) and entry['(Subtable)'] != processed_string:\n",
    "            running_t = False\n",
    "            running_p = False\n",
    "            running_i = False\n",
    "            running_lambda = False\n",
    "\n",
    "            z = entry['(Subtable)'].replace('\\'','\"').replace('\\’','\"').replace('’','\"').replace('‘','\"')            \n",
    "            subtable_as_list = ast.literal_eval(z)\n",
    "            \n",
    "            print(block.number)\n",
    "            variation_header = subtable_as_list[0]\n",
    "            print(variation_header)\n",
    "\n",
    "            # check headers to see which type of subtable this is.\n",
    "            running_t, running_p, running_i, P_units = check_subtable_type(variation_header)\n",
    "            \n",
    "            # iterate through the rows of the subtable\n",
    "            x_var_list, end_dict = process_subtable(subtable_as_list)\n",
    "            \n",
    "            # Now, loop everything together\n",
    "            print(x_var_list)\n",
    "            print(end_dict + \"\\n\")\n",
    "            end_list = list(end_dict.items())\n",
    "            \n",
    "            # Clean subtable tag. this needs to go first\n",
    "            block.entries[count]['(Subtable)'] = processed_string\n",
    "            \n",
    "            # Case A: just 1 temp list\n",
    "            if len(x_var_list) == 1:\n",
    "                for pka_list in end_list: # e.g. [(pK1, [3.1, 3.2, 3.3]), (pk2, [1.2, 2.2, 3.2])]\n",
    "                    for ct_x, x in enumerate(x_var_list[0]): #e.g. [5, 10, 15]\n",
    "                        entry_duplicate = copy.deepcopy(entry)\n",
    "                        entry_duplicate = assign_remarks(entry_duplicate, x, running_t, running_p, running_i, P_units)\n",
    "                        if pka_list[0] != 'lambda_0':\n",
    "                            entry_duplicate['pKa type'] = pka_list[0]\n",
    "                            entry_duplicate['pKa'] = pka_list[1][ct_x]\n",
    "                            if block.number == 2163: # This is a special case we need to remove.\n",
    "                                entry_duplicate['Remarks'] = entry_duplicate['Remarks'].replace('lambda_0=393.23','')\n",
    "                                entry_duplicate['Remarks'] += '\\nlambda_0={}'.format(lambdas[ct_x])\n",
    "                            block.entries.append(entry_duplicate)\n",
    "            \n",
    "            # Case B: multiple temp lists. Assume temp list 1 = first pKa, temp list 2 = second pKa\n",
    "            elif len(x_var_list) > 1:\n",
    "                for ct_xlist, x_list in enumerate(x_var_list): # e.g. [[5, 10, 15], [5, 15, 20, 25]]:\n",
    "                    for ct_x, x in enumerate(x_list): #e.g. [5, 10, 15]\n",
    "                        entry_duplicate = copy.deepcopy(entry)\n",
    "                        entry_duplicate = assign_remarks(entry_duplicate, x, running_t, running_p, running_i)\n",
    "                        entry_duplicate['pKa type'] = end_list[ct_xlist][0]\n",
    "                        entry_duplicate['pKa'] = end_list[ct_xlist][1][ct_x]\n",
    "                        block.entries.append(entry_duplicate)\n",
    "            \n",
    "            else:\n",
    "                raise ValueError(\"length of x_var_list = 0? (Error)\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a07f750b",
   "metadata": {},
   "source": [
    "# Tool for exporting names for analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e9f2a486",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry #</th>\n",
       "      <th>Name 1</th>\n",
       "      <th>Nickname 1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2004</td>\n",
       "      <td>Methanol</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2005</td>\n",
       "      <td>Methyl hydroperoxide</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2006</td>\n",
       "      <td>Methanethiol</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2007</td>\n",
       "      <td>methane, trinitro-</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2008</td>\n",
       "      <td>methane, dinitro-</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2009</td>\n",
       "      <td>methane, nitro-</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2010</td>\n",
       "      <td>Methanohydroxamic acid</td>\n",
       "      <td>Formohydroxamic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2011</td>\n",
       "      <td>methanedithioic acid, amino-</td>\n",
       "      <td>Dithiocarbamic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2012</td>\n",
       "      <td>methanesulfinic acid, hydroxy-</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Entry #                          Name 1            Nickname 1\n",
       "0     2004                        Methanol                      \n",
       "1     2005            Methyl hydroperoxide                      \n",
       "2     2006                    Methanethiol                      \n",
       "3     2007              methane, trinitro-                      \n",
       "4     2008               methane, dinitro-                      \n",
       "5     2009                 methane, nitro-                      \n",
       "6     2010          Methanohydroxamic acid  Formohydroxamic acid\n",
       "7     2011    methanedithioic acid, amino-   Dithiocarbamic acid\n",
       "8     2012  methanesulfinic acid, hydroxy-                      "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "block_nums = []\n",
    "names = [[],[]]\n",
    "nicknames = [[],[],[]]\n",
    "\n",
    "# Sort the blocks by numerical order and then export to DF for further processing their names.\n",
    "for block in sorted_blocklist:\n",
    "    block_nums.append(block.number)\n",
    "    for i in range(2):\n",
    "        try:\n",
    "            names[i].append(block.names[i])\n",
    "        except IndexError:\n",
    "            names[i].append('')\n",
    "#    for i in range(3):\n",
    "    for i in range(1):\n",
    "        try:\n",
    "            nicknames[i].append(block.nicknames[i])\n",
    "        except IndexError:\n",
    "            nicknames[i].append('')\n",
    "df = pd.DataFrame.from_dict({'Entry #': block_nums, 'Name 1':names[0], #'Name 2': names[1], \n",
    "                             'Nickname 1': nicknames[0], \n",
    "                            # 'Nickname 2': nicknames[1], 'Nickname 3': nicknames[2]\n",
    "                            })\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "fce1ef95",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"sample_names.csv\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a34e038",
   "metadata": {},
   "source": [
    "# Tool for assigning SMILES to DataEntryBlocks, need RDkit for this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "25998abc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_from_pubchempy(name : str) -> str:\n",
    "    \"\"\"\n",
    "    Query PubChem to obtain its isomeric SMILES string\n",
    "    \"\"\"\n",
    "    if my_isnull(name):\n",
    "        return None\n",
    "\n",
    "    name = str(name)\n",
    "\n",
    "    time.sleep(0.5)\n",
    "    name_set = set()\n",
    "    results = pcp.get_compounds(name, 'name')\n",
    "    for compound in results:\n",
    "        name_set.add(compound.isomeric_smiles)\n",
    "    \n",
    "    if len(name_set) == 1:\n",
    "        return list(name_set)[0]\n",
    "    else:\n",
    "        return None\n",
    "    \n",
    "def get_from_cirpy(name : str) -> str:\n",
    "    \"\"\"\n",
    "    Query cirpy to search CIR for SMILES\n",
    "    \"\"\"\n",
    "    if my_isnull(name):\n",
    "        return None\n",
    "\n",
    "    name = str(name)\n",
    "    time.sleep(0.5)\n",
    "    try:\n",
    "        smiles = cirpy.resolve(name, 'smiles', ['name_by_cir'])\n",
    "    except HTTPError:\n",
    "        smiles = ''\n",
    "    return smiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "e5983ecb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry #</th>\n",
       "      <th>Name 1</th>\n",
       "      <th>Nickname 1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2004</td>\n",
       "      <td>Methanol</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2005</td>\n",
       "      <td>Methyl hydroperoxide</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2006</td>\n",
       "      <td>Methanethiol</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2007</td>\n",
       "      <td>methane, trinitro-</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2008</td>\n",
       "      <td>methane, dinitro-</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2009</td>\n",
       "      <td>methane, nitro-</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2010</td>\n",
       "      <td>Methanohydroxamic acid</td>\n",
       "      <td>Formohydroxamic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2011</td>\n",
       "      <td>methanedithioic acid, amino-</td>\n",
       "      <td>Dithiocarbamic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2012</td>\n",
       "      <td>methanesulfinic acid, hydroxy-</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Entry #                          Name 1            Nickname 1\n",
       "0     2004                        Methanol                      \n",
       "1     2005            Methyl hydroperoxide                      \n",
       "2     2006                    Methanethiol                      \n",
       "3     2007              methane, trinitro-                      \n",
       "4     2008               methane, dinitro-                      \n",
       "5     2009                 methane, nitro-                      \n",
       "6     2010          Methanohydroxamic acid  Formohydroxamic acid\n",
       "7     2011    methanedithioic acid, amino-   Dithiocarbamic acid\n",
       "8     2012  methanesulfinic acid, hydroxy-                      "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36697d9a",
   "metadata": {},
   "source": [
    "### Create 'converged_smiles'\n",
    "\n",
    "Among the cheminformatics methods that could try to convert name to structure, see which ones agree."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ace3289-4948-4033-90e1-17e8e8c388e9",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "3c052c52-f04f-4b98-946b-09443327a830",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Resolving Name 1...\n",
      "Resolving nickname 1...\n"
     ]
    }
   ],
   "source": [
    "if 'opsin_name1' not in df:\n",
    "    print(\"Resolving Name 1...\")\n",
    "    df['opsin_name1'] = py2opsin(df['Name 1'].values)\n",
    "\n",
    "if 'opsin_nickname1' not in df:\n",
    "    print(\"Resolving nickname 1...\")\n",
    "    df['opsin_nickname1'] = py2opsin(df['Nickname 1'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f001b2cd-b3cc-4b7e-a6cd-430f5efe8d70",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "4426da25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Resolving Name 1...\n",
      "Resolving nickname 1...\n"
     ]
    }
   ],
   "source": [
    "if 'pubchem_name1' not in df:\n",
    "    print(\"Resolving Name 1...\")\n",
    "    df['pubchem_name1'] = df['Name 1'].apply(lambda x: get_from_pubchempy(x))\n",
    "\n",
    "if 'pubchem_nickname1' not in df:\n",
    "    print(\"Resolving nickname 1...\")\n",
    "    df['pubchem_nickname1'] = df['Nickname 1'].apply(lambda x: get_from_pubchempy(x))    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "9fb6858e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Resolving Name1...\n",
      "Resolving Nickname1...\n"
     ]
    }
   ],
   "source": [
    "if 'cirpy_name1' not in df:\n",
    "    print(\"Resolving Name1...\")\n",
    "    df['cirpy_name1'] = df['Name 1'].apply(lambda x: get_from_cirpy(x))\n",
    "\n",
    "if 'cirpy_nickname1' not in df:\n",
    "    print(\"Resolving Nickname1...\")\n",
    "    df['cirpy_nickname1'] = df['Nickname 1'].apply(lambda x: get_from_cirpy(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "48e08f42",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"sample_names_OUT.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "40fc7450",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error with SMILES set: {'O=CNO', 'NC(=O)O'} | molecule Methanohydroxamic acid\n"
     ]
    }
   ],
   "source": [
    "headers = ['opsin_name1', \n",
    "           'opsin_nickname1', \n",
    "           'cirpy_name1',\n",
    "           'cirpy_nickname1',\n",
    "           'pubchem_name1',\n",
    "           'pubchem_nickname1',\n",
    "          ]\n",
    "\n",
    "# Keep track of how many entries we have that are inconsistent, or empty because we couldn't translate it,\n",
    "# and of course how many were successful.\n",
    "inconsistent_ct = 0\n",
    "missing_ct = 0\n",
    "converged_ct = 0\n",
    "\n",
    "for ct, row in enumerate(df.iterrows()):\n",
    "    contributors = []\n",
    "    smiles_list = []\n",
    "    for header in headers:\n",
    "        smiles = row[1][header]\n",
    "        # If we have a valid SMILES, convert it to its canonical form.\n",
    "        if not pd.isnull(smiles) and smiles != '':\n",
    "            try:\n",
    "                mol = Chem.MolFromSmiles(smiles)\n",
    "                smiles_list.append(Chem.MolToSmiles(mol,isomericSmiles=True))                \n",
    "                contributors.append(header)\n",
    "            except SystemError:\n",
    "                pass\n",
    "            \n",
    "    # compare with manual SMILES also\n",
    "    smiles_set = set(smiles_list)\n",
    "\n",
    "    # Now, we look at \"smiles_set\" which is the set of unique smiles strings.\n",
    "    # If we have 1, then we have convergence.\n",
    "    # If we have more than 1, then we need to carefully examine them.\n",
    "    if len(smiles_set) > 1:\n",
    "        # Algorithm: Let's say there are only 2 entries. And one of them is the isomeric version.\n",
    "        # In that case, just append the isomeric version.\n",
    "        if len(smiles_set) == 2:\n",
    "            smiles_set_unwrapped = list(smiles_set)\n",
    "            mol0 = Chem.MolFromSmiles(smiles_set_unwrapped[0])\n",
    "            mol1 = Chem.MolFromSmiles(smiles_set_unwrapped[1])\n",
    "            if Chem.MolToSmiles(mol0, isomericSmiles=False) == Chem.MolToSmiles(mol1):\n",
    "                converged_smiles = smiles_set_unwrapped[0]\n",
    "                print(\"Isomeric: {} for molecule {}\".format(converged_smiles, row[1]['Name 1']))\n",
    "                df.loc[ct, 'type'] = 'isomeric'\n",
    "                converged_ct += 1\n",
    "            elif Chem.MolToSmiles(mol1, isomericSmiles=False) == Chem.MolToSmiles(mol0):\n",
    "                converged_smiles = smiles_set_unwrapped[1]\n",
    "                print(\"Isomeric: {} for molecule {}\".format(converged_smiles, row[1]['Name 1']))\n",
    "                df.loc[ct, 'type'] = 'isomeric'\n",
    "                converged_ct += 1\n",
    "            else:\n",
    "                converged_smiles = ''\n",
    "                inconsistent_ct += 1\n",
    "                print(\"error with SMILES set: {} | molecule {}\".format(smiles_set, row[1]['Name 1']))\n",
    "                df.loc[ct, 'type'] = 'inconsistent'\n",
    "        else: # 3 or more inconsistent...        \n",
    "            converged_smiles = ''\n",
    "            df.loc[ct, 'type'] = 'inconsistent'\n",
    "            print(\"error with SMILES set: {} | molecule {}\".format(smiles_set, row[1]['Name 1']))\n",
    "            inconsistent_ct += 1\n",
    "    elif len(smiles_set) == 0:\n",
    "        df.loc[ct, 'type'] = 'missing'\n",
    "        converged_smiles = ''\n",
    "        missing_ct += 1\n",
    "    else:\n",
    "        df.loc[ct, 'type'] = 'converged'\n",
    "        converged_smiles = smiles_set.pop()\n",
    "        converged_ct += 1\n",
    "        \n",
    "    df.loc[ct,'converged_smiles'] = converged_smiles\n",
    "    \n",
    "    # save the \"contributors\" which tells us which columns contribute to this convergence.\n",
    "    if len(converged_smiles) > 0:\n",
    "        df.loc[ct, 'contributors'] = str(contributors)\n",
    "    else:\n",
    "        df.loc[ct, 'contributors'] = ''\n",
    "    df.loc[ct, 'num_contributors'] = int(len(contributors))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "7950f9ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "0\n",
      "8\n",
      "Inconsistent total: 0.1111111111111111\n",
      "Missing total: 0.0\n",
      "Converged total: 0.8888888888888888\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry #</th>\n",
       "      <th>Name 1</th>\n",
       "      <th>Nickname 1</th>\n",
       "      <th>opsin_name1</th>\n",
       "      <th>opsin_nickname1</th>\n",
       "      <th>pubchem_name1</th>\n",
       "      <th>pubchem_nickname1</th>\n",
       "      <th>cirpy_name1</th>\n",
       "      <th>cirpy_nickname1</th>\n",
       "      <th>type</th>\n",
       "      <th>converged_smiles</th>\n",
       "      <th>contributors</th>\n",
       "      <th>num_contributors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2004</td>\n",
       "      <td>Methanol</td>\n",
       "      <td></td>\n",
       "      <td>CO</td>\n",
       "      <td></td>\n",
       "      <td>CO</td>\n",
       "      <td>None</td>\n",
       "      <td>CO</td>\n",
       "      <td>None</td>\n",
       "      <td>converged</td>\n",
       "      <td>CO</td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2005</td>\n",
       "      <td>Methyl hydroperoxide</td>\n",
       "      <td></td>\n",
       "      <td>COO</td>\n",
       "      <td></td>\n",
       "      <td>COO</td>\n",
       "      <td>None</td>\n",
       "      <td>COO</td>\n",
       "      <td>None</td>\n",
       "      <td>converged</td>\n",
       "      <td>COO</td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2006</td>\n",
       "      <td>Methanethiol</td>\n",
       "      <td></td>\n",
       "      <td>CS</td>\n",
       "      <td></td>\n",
       "      <td>CS</td>\n",
       "      <td>None</td>\n",
       "      <td>CS</td>\n",
       "      <td>None</td>\n",
       "      <td>converged</td>\n",
       "      <td>CS</td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2007</td>\n",
       "      <td>methane, trinitro-</td>\n",
       "      <td></td>\n",
       "      <td>[N+](=O)([O-])C([N+](=O)[O-])[N+](=O)[O-]</td>\n",
       "      <td></td>\n",
       "      <td>C([N+](=O)[O-])([N+](=O)[O-])[N+](=O)[O-]</td>\n",
       "      <td>None</td>\n",
       "      <td>[O-][N+](=O)C([N+]([O-])=O)[N+]([O-])=O</td>\n",
       "      <td>None</td>\n",
       "      <td>converged</td>\n",
       "      <td>O=[N+]([O-])C([N+](=O)[O-])[N+](=O)[O-]</td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2008</td>\n",
       "      <td>methane, dinitro-</td>\n",
       "      <td></td>\n",
       "      <td>[N+](=O)([O-])C[N+](=O)[O-]</td>\n",
       "      <td></td>\n",
       "      <td>C([N+](=O)[O-])[N+](=O)[O-]</td>\n",
       "      <td>None</td>\n",
       "      <td>[O-][N+](=O)C[N+]([O-])=O</td>\n",
       "      <td>None</td>\n",
       "      <td>converged</td>\n",
       "      <td>O=[N+]([O-])C[N+](=O)[O-]</td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2009</td>\n",
       "      <td>methane, nitro-</td>\n",
       "      <td></td>\n",
       "      <td>[N+](=O)([O-])C</td>\n",
       "      <td></td>\n",
       "      <td>C[N+](=O)[O-]</td>\n",
       "      <td>None</td>\n",
       "      <td>C[N+]([O-])=O</td>\n",
       "      <td>None</td>\n",
       "      <td>converged</td>\n",
       "      <td>C[N+](=O)[O-]</td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2010</td>\n",
       "      <td>Methanohydroxamic acid</td>\n",
       "      <td>Formohydroxamic acid</td>\n",
       "      <td>C(=O)NO</td>\n",
       "      <td>C(=O)NO</td>\n",
       "      <td>None</td>\n",
       "      <td>C(=O)NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NC(O)=O</td>\n",
       "      <td>inconsistent</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2011</td>\n",
       "      <td>methanedithioic acid, amino-</td>\n",
       "      <td>Dithiocarbamic acid</td>\n",
       "      <td>NC(=S)S</td>\n",
       "      <td>C(N)(S)=S</td>\n",
       "      <td>None</td>\n",
       "      <td>C(=S)(N)S</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>converged</td>\n",
       "      <td>NC(=S)S</td>\n",
       "      <td>['opsin_name1', 'opsin_nickname1', 'pubchem_ni...</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2012</td>\n",
       "      <td>methanesulfinic acid, hydroxy-</td>\n",
       "      <td></td>\n",
       "      <td>OCS(=O)O</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>converged</td>\n",
       "      <td>O=S(O)CO</td>\n",
       "      <td>['opsin_name1']</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Entry #                          Name 1            Nickname 1  \\\n",
       "0     2004                        Methanol                         \n",
       "1     2005            Methyl hydroperoxide                         \n",
       "2     2006                    Methanethiol                         \n",
       "3     2007              methane, trinitro-                         \n",
       "4     2008               methane, dinitro-                         \n",
       "5     2009                 methane, nitro-                         \n",
       "6     2010          Methanohydroxamic acid  Formohydroxamic acid   \n",
       "7     2011    methanedithioic acid, amino-   Dithiocarbamic acid   \n",
       "8     2012  methanesulfinic acid, hydroxy-                         \n",
       "\n",
       "                                 opsin_name1 opsin_nickname1  \\\n",
       "0                                         CO                   \n",
       "1                                        COO                   \n",
       "2                                         CS                   \n",
       "3  [N+](=O)([O-])C([N+](=O)[O-])[N+](=O)[O-]                   \n",
       "4                [N+](=O)([O-])C[N+](=O)[O-]                   \n",
       "5                            [N+](=O)([O-])C                   \n",
       "6                                    C(=O)NO         C(=O)NO   \n",
       "7                                    NC(=S)S       C(N)(S)=S   \n",
       "8                                   OCS(=O)O                   \n",
       "\n",
       "                               pubchem_name1 pubchem_nickname1  \\\n",
       "0                                         CO              None   \n",
       "1                                        COO              None   \n",
       "2                                         CS              None   \n",
       "3  C([N+](=O)[O-])([N+](=O)[O-])[N+](=O)[O-]              None   \n",
       "4                C([N+](=O)[O-])[N+](=O)[O-]              None   \n",
       "5                              C[N+](=O)[O-]              None   \n",
       "6                                       None           C(=O)NO   \n",
       "7                                       None         C(=S)(N)S   \n",
       "8                                       None              None   \n",
       "\n",
       "                               cirpy_name1 cirpy_nickname1          type  \\\n",
       "0                                       CO            None     converged   \n",
       "1                                      COO            None     converged   \n",
       "2                                       CS            None     converged   \n",
       "3  [O-][N+](=O)C([N+]([O-])=O)[N+]([O-])=O            None     converged   \n",
       "4                [O-][N+](=O)C[N+]([O-])=O            None     converged   \n",
       "5                            C[N+]([O-])=O            None     converged   \n",
       "6                                     None         NC(O)=O  inconsistent   \n",
       "7                                     None            None     converged   \n",
       "8                                     None            None     converged   \n",
       "\n",
       "                          converged_smiles  \\\n",
       "0                                       CO   \n",
       "1                                      COO   \n",
       "2                                       CS   \n",
       "3  O=[N+]([O-])C([N+](=O)[O-])[N+](=O)[O-]   \n",
       "4                O=[N+]([O-])C[N+](=O)[O-]   \n",
       "5                            C[N+](=O)[O-]   \n",
       "6                                            \n",
       "7                                  NC(=S)S   \n",
       "8                                 O=S(O)CO   \n",
       "\n",
       "                                        contributors  num_contributors  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']               3.0  \n",
       "1    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']               3.0  \n",
       "2    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']               3.0  \n",
       "3    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']               3.0  \n",
       "4    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']               3.0  \n",
       "5    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']               3.0  \n",
       "6                                                                  4.0  \n",
       "7  ['opsin_name1', 'opsin_nickname1', 'pubchem_ni...               3.0  \n",
       "8                                    ['opsin_name1']               1.0  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(inconsistent_ct)\n",
    "print(missing_ct)\n",
    "print(converged_ct)\n",
    "print(\"Inconsistent total: {}\".format(inconsistent_ct / (converged_ct + missing_ct + inconsistent_ct)))\n",
    "print(\"Missing total: {}\".format(missing_ct / (converged_ct + missing_ct + inconsistent_ct)))\n",
    "print(\"Converged total: {}\".format(converged_ct / (converged_ct + missing_ct + inconsistent_ct)))\n",
    "df.head(99)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83764c6c",
   "metadata": {},
   "source": [
    "### Wrap in the SMILES translations into the block entries.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "afb7953f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "entry_dict = {}\n",
    "for row in df.iterrows():\n",
    "    smiles = row[1]['converged_smiles']\n",
    "    contributors = row[1]['contributors']\n",
    "    num_contributors = row[1]['num_contributors']\n",
    "    if not my_isnull(smiles):\n",
    "        entry_dict[row[1]['Entry #']] = [smiles, contributors, num_contributors]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "10984ba9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{2004: ['CO', \"['opsin_name1', 'cirpy_name1', 'pubchem_name1']\", 3.0],\n",
       " 2005: ['COO', \"['opsin_name1', 'cirpy_name1', 'pubchem_name1']\", 3.0],\n",
       " 2006: ['CS', \"['opsin_name1', 'cirpy_name1', 'pubchem_name1']\", 3.0],\n",
       " 2007: ['O=[N+]([O-])C([N+](=O)[O-])[N+](=O)[O-]',\n",
       "  \"['opsin_name1', 'cirpy_name1', 'pubchem_name1']\",\n",
       "  3.0],\n",
       " 2008: ['O=[N+]([O-])C[N+](=O)[O-]',\n",
       "  \"['opsin_name1', 'cirpy_name1', 'pubchem_name1']\",\n",
       "  3.0],\n",
       " 2009: ['C[N+](=O)[O-]',\n",
       "  \"['opsin_name1', 'cirpy_name1', 'pubchem_name1']\",\n",
       "  3.0],\n",
       " 2011: ['NC(=S)S',\n",
       "  \"['opsin_name1', 'opsin_nickname1', 'pubchem_nickname1']\",\n",
       "  3.0],\n",
       " 2012: ['O=S(O)CO', \"['opsin_name1']\", 1.0]}"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entry_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d0132bba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CO\n",
      "COO\n",
      "CS\n",
      "O=[N+]([O-])C([N+](=O)[O-])[N+](=O)[O-]\n",
      "O=[N+]([O-])C[N+](=O)[O-]\n",
      "C[N+](=O)[O-]\n",
      "NC(=S)S\n",
      "O=S(O)CO\n"
     ]
    }
   ],
   "source": [
    "for block in sorted_blocklist:\n",
    "    try:\n",
    "        block.smiles = entry_dict[block.number][0]\n",
    "        block.contributors = entry_dict[block.number][1]\n",
    "        block.num_contributors = entry_dict[block.number][2]\n",
    "        print(block.smiles)\n",
    "    except KeyError:\n",
    "        block.smiles = None\n",
    "        block.contributors = None\n",
    "        block.num_contributors = None\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f7fe818",
   "metadata": {},
   "source": [
    "### Convert everything to a CSV."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "9da14f4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ref_remarks_wrapper(refs, entryref):\n",
    "    \"\"\"\n",
    "    Wrapper that takes the ref for an entry and applies it to the BLOCK's ref.\n",
    "    \"\"\"\n",
    "    if my_isnull(entryref):\n",
    "        return \"\"\n",
    "    else:\n",
    "        return refs[entryref]\n",
    "\n",
    "def unpack(strlist):\n",
    "    \"\"\"\n",
    "    Unpacks a list into a string separated by semicolons.\n",
    "    \"\"\"\n",
    "    if type(strlist) == list:\n",
    "        return_string = ''\n",
    "        for ct,string in enumerate(strlist):\n",
    "            if ct == 0:\n",
    "                return_string = string\n",
    "            else:\n",
    "                return_string += \"; \" + string\n",
    "        return return_string\n",
    "    else:\n",
    "        return strlist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "aeed3ab0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_187751/3094118832.py:8: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  \"pka_type\": entry[0],\n",
      "/tmp/ipykernel_187751/3094118832.py:9: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  \"pka_value\": entry[1],\n",
      "/tmp/ipykernel_187751/3094118832.py:10: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  \"T\": entry[2],\n",
      "/tmp/ipykernel_187751/3094118832.py:11: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  \"remarks\": entry[3],\n",
      "/tmp/ipykernel_187751/3094118832.py:12: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  \"method\": entry[4],\n",
      "/tmp/ipykernel_187751/3094118832.py:13: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  \"assessment\": entry[5],\n",
      "/tmp/ipykernel_187751/3094118832.py:14: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  \"ref\": entry[6],\n",
      "/tmp/ipykernel_187751/3094118832.py:15: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  \"ref_remarks\": unpack(ref_remarks_wrapper(block.refs, entry[6])),\n",
      "/tmp/ipykernel_187751/3094118832.py:5: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  df = pd.concat([df, pd.DataFrame({\n",
      "/tmp/ipykernel_187751/3094118832.py:8: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  \"pka_type\": entry[0],\n",
      "/tmp/ipykernel_187751/3094118832.py:9: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  \"pka_value\": entry[1],\n",
      "/tmp/ipykernel_187751/3094118832.py:10: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  \"T\": entry[2],\n",
      "/tmp/ipykernel_187751/3094118832.py:11: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  \"remarks\": entry[3],\n",
      "/tmp/ipykernel_187751/3094118832.py:12: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  \"method\": entry[4],\n",
      "/tmp/ipykernel_187751/3094118832.py:13: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  \"assessment\": entry[5],\n",
      "/tmp/ipykernel_187751/3094118832.py:14: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  \"ref\": entry[6],\n",
      "/tmp/ipykernel_187751/3094118832.py:15: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  \"ref_remarks\": unpack(ref_remarks_wrapper(block.refs, entry[6])),\n"
     ]
    }
   ],
   "source": [
    "df = pd.DataFrame(columns=['entry_#','SMILES','pka_type','pka_value','T','remarks','method','assessment','ref','ref_remarks','entry_remarks','original_IUPAC_names'])\n",
    "\n",
    "for block in sorted_blocklist:\n",
    "    for entry in block.entries:\n",
    "        df = pd.concat([df, pd.DataFrame({\n",
    "            \"entry_#\": block.number,\n",
    "            \"SMILES\": block.smiles,\n",
    "            \"pka_type\": entry[0],\n",
    "            \"pka_value\": entry[1],\n",
    "            \"T\": entry[2],\n",
    "            \"remarks\": entry[3],\n",
    "            \"method\": entry[4],\n",
    "            \"assessment\": entry[5],\n",
    "            \"ref\": entry[6],\n",
    "            \"ref_remarks\": unpack(ref_remarks_wrapper(block.refs, entry[6])),\n",
    "            \"entry_remarks\": block.description,\n",
    "            \"original_IUPAC_names\": block.names[0],\n",
    "            \"original_IUPAC_nicknames\": unpack(block.nicknames),\n",
    "            \"name_contributors\": block.contributors,\n",
    "            \"num_name_contributors\": block.num_contributors,\n",
    "        },index=[0])] )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7e1eb8b-8de7-41cc-a6a6-d961a9abfca0",
   "metadata": {},
   "source": [
    "Note that there is a small error that was introduced from the 1st notebook for entry 2007. There was an entry that should have pKa=0.02 but it was lumped into the block remarks.\n",
    "\n",
    "We manually examined the entries in the dataset for mistakes like these."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "a3c649f3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entry_#</th>\n",
       "      <th>SMILES</th>\n",
       "      <th>pka_type</th>\n",
       "      <th>pka_value</th>\n",
       "      <th>T</th>\n",
       "      <th>remarks</th>\n",
       "      <th>method</th>\n",
       "      <th>assessment</th>\n",
       "      <th>ref</th>\n",
       "      <th>ref_remarks</th>\n",
       "      <th>entry_remarks</th>\n",
       "      <th>original_IUPAC_names</th>\n",
       "      <th>original_IUPAC_nicknames</th>\n",
       "      <th>name_contributors</th>\n",
       "      <th>num_name_contributors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2004</td>\n",
       "      <td>CO</td>\n",
       "      <td>pKa</td>\n",
       "      <td>15.50</td>\n",
       "      <td>25.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>B8</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>Methanol</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2004</td>\n",
       "      <td>CO</td>\n",
       "      <td>pKa</td>\n",
       "      <td>15.09</td>\n",
       "      <td>25.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KIN</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>M126</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>Methanol</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2005</td>\n",
       "      <td>COO</td>\n",
       "      <td>pKa</td>\n",
       "      <td>11.50</td>\n",
       "      <td>20.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O5</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>E27</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>Methyl hydroperoxide</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2006</td>\n",
       "      <td>CS</td>\n",
       "      <td>pKa</td>\n",
       "      <td>10.33</td>\n",
       "      <td>25.0</td>\n",
       "      <td>1% ethanol, gas solubility method</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>K57</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>Methanethiol</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2007</td>\n",
       "      <td>O=[N+]([O-])C([N+](=O)[O-])[N+](=O)[O-]</td>\n",
       "      <td>pKa</td>\n",
       "      <td>0.14</td>\n",
       "      <td>20.0</td>\n",
       "      <td>In aqueous HClO4 Ho scale</td>\n",
       "      <td>O6</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>T58a</td>\n",
       "      <td></td>\n",
       "      <td>- 0.02  Thermodynamic quantities are derived ...</td>\n",
       "      <td>methane, trinitro-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2007</td>\n",
       "      <td>O=[N+]([O-])C([N+](=O)[O-])[N+](=O)[O-]</td>\n",
       "      <td>pKa</td>\n",
       "      <td>0.06</td>\n",
       "      <td>25.5</td>\n",
       "      <td>Mixed constant</td>\n",
       "      <td>O5</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>H5</td>\n",
       "      <td></td>\n",
       "      <td>- 0.02  Thermodynamic quantities are derived ...</td>\n",
       "      <td>methane, trinitro-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2007</td>\n",
       "      <td>O=[N+]([O-])C([N+](=O)[O-])[N+](=O)[O-]</td>\n",
       "      <td>pKa</td>\n",
       "      <td>0.05</td>\n",
       "      <td>9.6</td>\n",
       "      <td>Mixed constant</td>\n",
       "      <td>O5</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>H5</td>\n",
       "      <td></td>\n",
       "      <td>- 0.02  Thermodynamic quantities are derived ...</td>\n",
       "      <td>methane, trinitro-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2007</td>\n",
       "      <td>O=[N+]([O-])C([N+](=O)[O-])[N+](=O)[O-]</td>\n",
       "      <td>pKa</td>\n",
       "      <td>0.23</td>\n",
       "      <td>5.0</td>\n",
       "      <td>In aqueous HCl Ho scale</td>\n",
       "      <td>O6</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>N39, S82</td>\n",
       "      <td>Thermodynamic quantities are derived from the ...</td>\n",
       "      <td>- 0.02  Thermodynamic quantities are derived ...</td>\n",
       "      <td>methane, trinitro-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2007</td>\n",
       "      <td>O=[N+]([O-])C([N+](=O)[O-])[N+](=O)[O-]</td>\n",
       "      <td>pKa</td>\n",
       "      <td>0.17</td>\n",
       "      <td>20.0</td>\n",
       "      <td>In aqueous HCl Ho scale</td>\n",
       "      <td>O6</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>N39, S82</td>\n",
       "      <td>Thermodynamic quantities are derived from the ...</td>\n",
       "      <td>- 0.02  Thermodynamic quantities are derived ...</td>\n",
       "      <td>methane, trinitro-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2007</td>\n",
       "      <td>O=[N+]([O-])C([N+](=O)[O-])[N+](=O)[O-]</td>\n",
       "      <td>pKa</td>\n",
       "      <td>0.11</td>\n",
       "      <td>40.0</td>\n",
       "      <td>In aqueous HCl Ho scale</td>\n",
       "      <td>O6</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>N39, S82</td>\n",
       "      <td>Thermodynamic quantities are derived from the ...</td>\n",
       "      <td>- 0.02  Thermodynamic quantities are derived ...</td>\n",
       "      <td>methane, trinitro-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2008</td>\n",
       "      <td>O=[N+]([O-])C[N+](=O)[O-]</td>\n",
       "      <td>pKa</td>\n",
       "      <td>3.63</td>\n",
       "      <td>20.0</td>\n",
       "      <td>I=0.06</td>\n",
       "      <td>O5</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>N39</td>\n",
       "      <td>Thermodynamic quantities are derived from the ...</td>\n",
       "      <td>None</td>\n",
       "      <td>methane, dinitro-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2008</td>\n",
       "      <td>O=[N+]([O-])C[N+](=O)[O-]</td>\n",
       "      <td>pKa</td>\n",
       "      <td>3.72</td>\n",
       "      <td>5.0</td>\n",
       "      <td>I=0.06</td>\n",
       "      <td>O5</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>N39</td>\n",
       "      <td>Thermodynamic quantities are derived from the ...</td>\n",
       "      <td>None</td>\n",
       "      <td>methane, dinitro-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2008</td>\n",
       "      <td>O=[N+]([O-])C[N+](=O)[O-]</td>\n",
       "      <td>pKa</td>\n",
       "      <td>3.60</td>\n",
       "      <td>20.0</td>\n",
       "      <td>I=0.06</td>\n",
       "      <td>O5</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>N39</td>\n",
       "      <td>Thermodynamic quantities are derived from the ...</td>\n",
       "      <td>None</td>\n",
       "      <td>methane, dinitro-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2008</td>\n",
       "      <td>O=[N+]([O-])C[N+](=O)[O-]</td>\n",
       "      <td>pKa</td>\n",
       "      <td>3.51</td>\n",
       "      <td>40.0</td>\n",
       "      <td>I=0.06</td>\n",
       "      <td>O5</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>N39</td>\n",
       "      <td>Thermodynamic quantities are derived from the ...</td>\n",
       "      <td>None</td>\n",
       "      <td>methane, dinitro-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2008</td>\n",
       "      <td>O=[N+]([O-])C[N+](=O)[O-]</td>\n",
       "      <td>pKa</td>\n",
       "      <td>3.43</td>\n",
       "      <td>60.0</td>\n",
       "      <td>I=0.06</td>\n",
       "      <td>O5</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>N39</td>\n",
       "      <td>Thermodynamic quantities are derived from the ...</td>\n",
       "      <td>None</td>\n",
       "      <td>methane, dinitro-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2008</td>\n",
       "      <td>O=[N+]([O-])C[N+](=O)[O-]</td>\n",
       "      <td>pKa</td>\n",
       "      <td>3.57</td>\n",
       "      <td>25.0</td>\n",
       "      <td>I=0.06</td>\n",
       "      <td>O5</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>A7</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>methane, dinitro-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2009</td>\n",
       "      <td>C[N+](=O)[O-]</td>\n",
       "      <td>pKa</td>\n",
       "      <td>10.45</td>\n",
       "      <td>10.0</td>\n",
       "      <td>c=0.005-0.017, mixed constant</td>\n",
       "      <td>E3bg</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>T67</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>methane, nitro-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2009</td>\n",
       "      <td>C[N+](=O)[O-]</td>\n",
       "      <td>pKa</td>\n",
       "      <td>10.33</td>\n",
       "      <td>18.0</td>\n",
       "      <td>c=0.005-0.017, mixed constant</td>\n",
       "      <td>E3bg</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>T67</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>methane, nitro-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2009</td>\n",
       "      <td>C[N+](=O)[O-]</td>\n",
       "      <td>pKa</td>\n",
       "      <td>10.21</td>\n",
       "      <td>25.0</td>\n",
       "      <td>c=0.005-0.017, mixed constant</td>\n",
       "      <td>E3bg</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>T67</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>methane, nitro-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2009</td>\n",
       "      <td>C[N+](=O)[O-]</td>\n",
       "      <td>pKa</td>\n",
       "      <td>10.24</td>\n",
       "      <td>25.0</td>\n",
       "      <td>c=0.005-0.017, mixed constant</td>\n",
       "      <td>E3bg</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>W22</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>methane, nitro-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1', 'cirpy_name1', 'pubchem_name1']</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2010</td>\n",
       "      <td>None</td>\n",
       "      <td>pKa</td>\n",
       "      <td>8.65</td>\n",
       "      <td>NaN</td>\n",
       "      <td>I=0.2(NaCl)</td>\n",
       "      <td>E3bg</td>\n",
       "      <td>Uncert.</td>\n",
       "      <td>C72</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>Methanohydroxamic acid</td>\n",
       "      <td>Formohydroxamic acid</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2011</td>\n",
       "      <td>NC(=S)S</td>\n",
       "      <td>pKa</td>\n",
       "      <td>2.95</td>\n",
       "      <td>25.0</td>\n",
       "      <td>c=0.002-0.01</td>\n",
       "      <td>C2</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>G4</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>methanedithioic acid, amino-</td>\n",
       "      <td>Dithiocarbamic acid</td>\n",
       "      <td>['opsin_name1', 'opsin_nickname1', 'pubchem_ni...</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2012</td>\n",
       "      <td>O=S(O)CO</td>\n",
       "      <td>pKa</td>\n",
       "      <td>1.65</td>\n",
       "      <td>20.0</td>\n",
       "      <td>c=0.1, mixed constant</td>\n",
       "      <td>E3bg</td>\n",
       "      <td>Approx.</td>\n",
       "      <td>R54</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>methanesulfinic acid, hydroxy-</td>\n",
       "      <td></td>\n",
       "      <td>['opsin_name1']</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  entry_#                                   SMILES pka_type  pka_value     T  \\\n",
       "0    2004                                       CO      pKa      15.50  25.0   \n",
       "0    2004                                       CO      pKa      15.09  25.0   \n",
       "0    2005                                      COO      pKa      11.50  20.0   \n",
       "0    2006                                       CS      pKa      10.33  25.0   \n",
       "0    2007  O=[N+]([O-])C([N+](=O)[O-])[N+](=O)[O-]      pKa       0.14  20.0   \n",
       "0    2007  O=[N+]([O-])C([N+](=O)[O-])[N+](=O)[O-]      pKa       0.06  25.5   \n",
       "0    2007  O=[N+]([O-])C([N+](=O)[O-])[N+](=O)[O-]      pKa       0.05   9.6   \n",
       "0    2007  O=[N+]([O-])C([N+](=O)[O-])[N+](=O)[O-]      pKa       0.23   5.0   \n",
       "0    2007  O=[N+]([O-])C([N+](=O)[O-])[N+](=O)[O-]      pKa       0.17  20.0   \n",
       "0    2007  O=[N+]([O-])C([N+](=O)[O-])[N+](=O)[O-]      pKa       0.11  40.0   \n",
       "0    2008                O=[N+]([O-])C[N+](=O)[O-]      pKa       3.63  20.0   \n",
       "0    2008                O=[N+]([O-])C[N+](=O)[O-]      pKa       3.72   5.0   \n",
       "0    2008                O=[N+]([O-])C[N+](=O)[O-]      pKa       3.60  20.0   \n",
       "0    2008                O=[N+]([O-])C[N+](=O)[O-]      pKa       3.51  40.0   \n",
       "0    2008                O=[N+]([O-])C[N+](=O)[O-]      pKa       3.43  60.0   \n",
       "0    2008                O=[N+]([O-])C[N+](=O)[O-]      pKa       3.57  25.0   \n",
       "0    2009                            C[N+](=O)[O-]      pKa      10.45  10.0   \n",
       "0    2009                            C[N+](=O)[O-]      pKa      10.33  18.0   \n",
       "0    2009                            C[N+](=O)[O-]      pKa      10.21  25.0   \n",
       "0    2009                            C[N+](=O)[O-]      pKa      10.24  25.0   \n",
       "0    2010                                     None      pKa       8.65   NaN   \n",
       "0    2011                                  NC(=S)S      pKa       2.95  25.0   \n",
       "0    2012                                 O=S(O)CO      pKa       1.65  20.0   \n",
       "\n",
       "                             remarks method assessment       ref  \\\n",
       "0                                NaN     C3    Uncert.        B8   \n",
       "0                                NaN    KIN    Uncert.      M126   \n",
       "0                                NaN     O5    Uncert.       E27   \n",
       "0  1% ethanol, gas solubility method    NaN    Uncert.       K57   \n",
       "0          In aqueous HClO4 Ho scale     O6    Uncert.      T58a   \n",
       "0                     Mixed constant     O5    Uncert.        H5   \n",
       "0                     Mixed constant     O5    Uncert.        H5   \n",
       "0            In aqueous HCl Ho scale     O6    Uncert.  N39, S82   \n",
       "0            In aqueous HCl Ho scale     O6    Uncert.  N39, S82   \n",
       "0            In aqueous HCl Ho scale     O6    Uncert.  N39, S82   \n",
       "0                             I=0.06     O5    Approx.       N39   \n",
       "0                             I=0.06     O5    Approx.       N39   \n",
       "0                             I=0.06     O5    Approx.       N39   \n",
       "0                             I=0.06     O5    Approx.       N39   \n",
       "0                             I=0.06     O5    Approx.       N39   \n",
       "0                             I=0.06     O5    Approx.        A7   \n",
       "0      c=0.005-0.017, mixed constant   E3bg    Approx.       T67   \n",
       "0      c=0.005-0.017, mixed constant   E3bg    Approx.       T67   \n",
       "0      c=0.005-0.017, mixed constant   E3bg    Approx.       T67   \n",
       "0      c=0.005-0.017, mixed constant   E3bg    Approx.       W22   \n",
       "0                        I=0.2(NaCl)   E3bg    Uncert.       C72   \n",
       "0                       c=0.002-0.01     C2    Approx.        G4   \n",
       "0              c=0.1, mixed constant   E3bg    Approx.       R54   \n",
       "\n",
       "                                         ref_remarks  \\\n",
       "0                                                      \n",
       "0                                                      \n",
       "0                                                      \n",
       "0                                                      \n",
       "0                                                      \n",
       "0                                                      \n",
       "0                                                      \n",
       "0  Thermodynamic quantities are derived from the ...   \n",
       "0  Thermodynamic quantities are derived from the ...   \n",
       "0  Thermodynamic quantities are derived from the ...   \n",
       "0  Thermodynamic quantities are derived from the ...   \n",
       "0  Thermodynamic quantities are derived from the ...   \n",
       "0  Thermodynamic quantities are derived from the ...   \n",
       "0  Thermodynamic quantities are derived from the ...   \n",
       "0  Thermodynamic quantities are derived from the ...   \n",
       "0                                                      \n",
       "0                                                      \n",
       "0                                                      \n",
       "0                                                      \n",
       "0                                                      \n",
       "0                                                      \n",
       "0                                                      \n",
       "0                                                      \n",
       "\n",
       "                                       entry_remarks  \\\n",
       "0                                               None   \n",
       "0                                               None   \n",
       "0                                               None   \n",
       "0                                               None   \n",
       "0   - 0.02  Thermodynamic quantities are derived ...   \n",
       "0   - 0.02  Thermodynamic quantities are derived ...   \n",
       "0   - 0.02  Thermodynamic quantities are derived ...   \n",
       "0   - 0.02  Thermodynamic quantities are derived ...   \n",
       "0   - 0.02  Thermodynamic quantities are derived ...   \n",
       "0   - 0.02  Thermodynamic quantities are derived ...   \n",
       "0                                               None   \n",
       "0                                               None   \n",
       "0                                               None   \n",
       "0                                               None   \n",
       "0                                               None   \n",
       "0                                               None   \n",
       "0                                               None   \n",
       "0                                               None   \n",
       "0                                               None   \n",
       "0                                               None   \n",
       "0                                               None   \n",
       "0                                               None   \n",
       "0                                               None   \n",
       "\n",
       "             original_IUPAC_names original_IUPAC_nicknames  \\\n",
       "0                        Methanol                            \n",
       "0                        Methanol                            \n",
       "0            Methyl hydroperoxide                            \n",
       "0                    Methanethiol                            \n",
       "0              methane, trinitro-                            \n",
       "0              methane, trinitro-                            \n",
       "0              methane, trinitro-                            \n",
       "0              methane, trinitro-                            \n",
       "0              methane, trinitro-                            \n",
       "0              methane, trinitro-                            \n",
       "0               methane, dinitro-                            \n",
       "0               methane, dinitro-                            \n",
       "0               methane, dinitro-                            \n",
       "0               methane, dinitro-                            \n",
       "0               methane, dinitro-                            \n",
       "0               methane, dinitro-                            \n",
       "0                 methane, nitro-                            \n",
       "0                 methane, nitro-                            \n",
       "0                 methane, nitro-                            \n",
       "0                 methane, nitro-                            \n",
       "0          Methanohydroxamic acid     Formohydroxamic acid   \n",
       "0    methanedithioic acid, amino-      Dithiocarbamic acid   \n",
       "0  methanesulfinic acid, hydroxy-                            \n",
       "\n",
       "                                   name_contributors  num_name_contributors  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0    ['opsin_name1', 'cirpy_name1', 'pubchem_name1']                    3.0  \n",
       "0                                               None                    NaN  \n",
       "0  ['opsin_name1', 'opsin_nickname1', 'pubchem_ni...                    3.0  \n",
       "0                                    ['opsin_name1']                    1.0  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "33e86ccb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"sample_done.csv\",index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
